esp-hal/src/system.rs

```diff
@@ -145,19 +144,22 @@ impl Peripheral {
     }
 }
 
-static PERIPHERAL_REF_COUNT: [AtomicUsize; Peripheral::COUNT] =
-    [const { AtomicUsize::new(0) }; Peripheral::COUNT];
+static PERIPHERAL_REF_COUNT: Mutex<RefCell<[usize; Peripheral::COUNT]>> =
```

Comment:
The borrow checks can be skipped here by doing `Mutex<[Cell<usize>; Peripheral::COUNT]>` instead, I'll let you decided if it's worth the `.get()` and `.set(....)`.